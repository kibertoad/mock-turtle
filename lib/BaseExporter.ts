import mkdirp from 'mkdirp-promise'
import { validationHelper as validate } from 'validation-utils'
import path from 'path'
import fs from 'fs'

export interface Response {
  body: any
  header: any
  status: number
  text: string
}

export interface ExporterConfig {
  doNotOverwrite: boolean
}

export abstract class BaseExporter {
  protected readonly pathToMocks: string
  private readonly extension: string
  private readonly doNotOverwrite: boolean

  public constructor(pathToMocks: string, extension: string, config: ExporterConfig) {
    validate.string(pathToMocks, 'pathToFixture is mandatory and must be a string')
    this.pathToMocks = pathToMocks
    this.extension = extension
    this.doNotOverwrite = config.doNotOverwrite
  }

  protected resolveFixturePath(name: string): string {
    const pathToMock = path.resolve(this.pathToMocks, `${name}.${this.extension}`)

    let resolvedPathToMock = pathToMock
    let counter = 0
    if (this.doNotOverwrite) {
      while (fs.existsSync(resolvedPathToMock)) {
        resolvedPathToMock = path.resolve(
          this.pathToMocks,
          `${name}_${++counter}.${this.extension}`
        )
      }
    }

    return resolvedPathToMock
  }

  protected async saveFixture(mockName: string, responseMock: string) {
    await mkdirp(this.pathToMocks)
    const path = this.resolveFixturePath(mockName)
    fs.writeFileSync(path, responseMock)
  }
}
