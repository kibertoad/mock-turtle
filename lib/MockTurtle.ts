import nockNamespace, {
  HttpHeaders,
  POJO,
  ReplyBody,
  ReplyCallbackResult,
  RequestBodyMatcher
} from 'nock'
import { Url } from 'url'
import deepMerge from 'deepmerge'

export interface GlobalOptions {
  basePath: string | RegExp | Url
  nockOptions?: nockNamespace.Options
  allowProtocolOmission?: boolean
}

export interface EndpointOptions {
  anyParams?: boolean
  requestQuery?: nockNamespace.RequestBodyMatcher
  requestBody?: nockNamespace.RequestBodyMatcher
  delayConnection?: number
  times?: number
  errorMessage?: string | POJO
}

export type ResponseFunction = (path: string, requestBody: POJO) => ReplyCallbackResult

export type EndpointResponse =
  | {
      responseCode: number
      body?: ReplyBody
      headers?: HttpHeaders
    }
  | POJO
  | ResponseFunction

enum HttpMethod {
  get = 'get',
  post = 'post'
}

type Nock = (
  basePath: string | RegExp | Url,
  options?: nockNamespace.Options
) => nockNamespace.Scope

export class MockTurtle {
  private readonly nock: Nock
  private readonly optionDefaults: Partial<GlobalOptions>

  public constructor(nock: Nock, optionDefaults?: GlobalOptions) {
    this.nock = nock
    this.optionDefaults = optionDefaults || {
      nockOptions: {}
    }
  }

  public mockGet(
    endpointPath: string | RegExp | ((uri: string) => boolean),
    mockedResponse?: EndpointResponse,
    endpointOptions?: EndpointOptions,
    optionOverrides?: Partial<GlobalOptions>
  ): nockNamespace.Interceptor {
    return this.mockEndpoint(
      HttpMethod.get,
      endpointPath,
      endpointOptions,
      mockedResponse,
      optionOverrides
    )
  }

  public mockPost(
    endpointPath: string | RegExp | ((uri: string) => boolean),
    mockedResponse?: EndpointResponse,
    endpointOptions?: EndpointOptions,
    optionOverrides?: Partial<GlobalOptions>
  ): nockNamespace.Interceptor {
    return this.mockEndpoint(
      HttpMethod.post,
      endpointPath,
      endpointOptions,
      mockedResponse,
      optionOverrides
    )
  }

  public disableExternalCalls() {
    // @ts-ignore
    this.nock.disableNetConnect()
  }

  public reset() {
    // @ts-ignore
    this.nock.cleanAll()
    // @ts-ignore
    this.nock.enableNetConnect()
  }

  private mockEndpoint(
    method: HttpMethod,
    endpointPath: string | RegExp | ((uri: string) => boolean),
    endpointOptions?: EndpointOptions,
    mockedResponse?: EndpointResponse,
    optionOverrides?: Partial<GlobalOptions>
  ): nockNamespace.Interceptor {
    endpointPath = endpointPath || '/'

    const resolvedGlobalOptions: GlobalOptions = deepMerge(
      this.optionDefaults,
      optionOverrides || {}
    )
    const resolvedEndpointOptions: EndpointOptions = endpointOptions || {}

    validateOptions(resolvedGlobalOptions, resolvedEndpointOptions)
    let bodyParam: RequestBodyMatcher | undefined
    if (resolvedEndpointOptions.requestBody) {
      bodyParam = resolvedEndpointOptions.requestBody
    }
    if (resolvedEndpointOptions.anyParams) {
      bodyParam = () => true
    }

    const mockBuilder = this.nock(
      resolvedGlobalOptions.basePath,
      resolvedGlobalOptions.nockOptions
    )[method](endpointPath, bodyParam)

    if (resolvedEndpointOptions.requestQuery) {
      mockBuilder.query(resolvedEndpointOptions.requestQuery)
    }
    if (resolvedEndpointOptions.anyParams) {
      mockBuilder.query(() => true)
    }

    if (resolvedEndpointOptions.delayConnection) {
      mockBuilder.delayConnection(resolvedEndpointOptions.delayConnection)
    }

    if (resolvedEndpointOptions.times) {
      mockBuilder.times(resolvedEndpointOptions.times)
    }

    if (resolvedEndpointOptions.errorMessage) {
      mockBuilder.replyWithError(resolvedEndpointOptions.errorMessage)
    }

    setReply(mockBuilder, mockedResponse)
    return mockBuilder
  }
}

function validateOptions(
  resolvedGlobalOptions: GlobalOptions,
  resolvedEndpointOptions: EndpointOptions
) {
  if (
    !resolvedGlobalOptions.allowProtocolOmission &&
    typeof resolvedGlobalOptions.basePath === 'string' &&
    !resolvedGlobalOptions.basePath.startsWith('http')
  ) {
    throw new Error(
      `Please add "http://" or "https://" to your baseUrl, otherwise mocking is unlikely to work. If you are pretty sure you don't need it, set "allowProtocolOmission" to true.`
    )
  }

  if (
    resolvedEndpointOptions.anyParams &&
    (resolvedEndpointOptions.requestBody || resolvedEndpointOptions.requestQuery)
  ) {
    throw new Error(
      'When "anyParams" flag is set to true, no "requestBody" or "requestQuery" can be set.'
    )
  }
}

function setReply(mockBuilder: nockNamespace.Interceptor, endpointReply?: EndpointResponse) {
  // Handle reply callback function
  if (typeof endpointReply === 'function') {
    mockBuilder.reply(endpointReply as any)
    return
  }

  // Handle static reply
  const resolvedReply: EndpointResponse = endpointReply
    ? resolveStaticResponse(endpointReply)
    : {
        responseCode: 200
      }

  mockBuilder.reply(resolvedReply.responseCode, resolvedReply.body, resolvedReply.headers)
}

function resolveStaticResponse(response: POJO): EndpointResponse {
  if (response.responseCode) {
    return response
  }
  return {
    responseCode: 200,
    body: response
  }
}
