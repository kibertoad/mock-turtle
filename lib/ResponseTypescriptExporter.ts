import { BaseExporter, ExporterConfig, Response } from './BaseExporter'
import { EndpointResponse } from './MockTurtle'

const DEFAULT_CONFIG: ExporterConfig = Object.freeze({
  doNotOverwrite: true
})

export class ResponseTypescriptExporter extends BaseExporter {
  public constructor(pathToMocks: string, config?: ExporterConfig) {
    super(pathToMocks, 'ts', config || DEFAULT_CONFIG)
  }

  public async exportResponseBody(mockName: string, response: Response) {
    const mock = `export const ResponseMock = Object.freeze(
  ${JSON.stringify(response.body, null, 4)}
)`
    await this.saveFixture(mockName, mock)
  }

  public async exportFullResponse(mockName: string, response: Response) {
    const responseMock: EndpointResponse = {
      body: response.body,
      headers: response.header,
      responseCode: response.status
    }

    const mock = `export const ResponseMock = Object.freeze(
  ${JSON.stringify(responseMock, null, 4)}
)`
    await this.saveFixture(mockName, mock)
  }
}
