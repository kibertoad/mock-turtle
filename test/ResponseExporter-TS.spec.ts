import request from 'superagent'
import { initHelper } from './utils/nockbackHelper'
import { ResponseTypescriptExporter } from '../lib/ResponseTypescriptExporter'
import rimraf from 'rimraf'

const exporter = new ResponseTypescriptExporter(__dirname + '/nock-responses')

describe('ResponseExporter', () => {
  describe('Export to TypeScript', () => {
    it('Export only body', async () => {
      rimraf.sync(getMockPath('todoMock.ts'))
      const helper = initHelper(__dirname)

      await helper.nockBack('placeholder-GET.json', async () => {
        const response = await request.get('https://jsonplaceholder.typicode.com/todos/1').set({
          'Accept-Encoding': 'identity' // This will prevent responses from being compressed which is preferred for mock recording
        })
        await exporter.exportResponseBody('todoMock', response)
        const importedMock = await import(getMockPath('todoMock.ts'))
        expect(importedMock).toMatchSnapshot()
      })
    })

    it('Export full response', async () => {
      rimraf.sync(getMockPath('todoFullMock.ts'))
      const helper = initHelper(__dirname)

      await helper.nockBack('placeholder-GET.json', async () => {
        const response = await request.get('https://jsonplaceholder.typicode.com/todos/1').set({
          'Accept-Encoding': 'identity' // This will prevent responses from being compressed which is preferred for mock recording
        })
        await exporter.exportFullResponse('todoFullMock', response)
        const importedMock = await import(getMockPath('todoFullMock.ts'))
        expect(importedMock).toMatchSnapshot()
      })
    })

    it('Does not overwrite mocks by default', async () => {
      rimraf.sync(getMockPath('todoMock.ts'))
      rimraf.sync(getMockPath('todoMock_1.ts'))
      const helper = initHelper(__dirname)

      await helper.nockBack('placeholder-GET.json', async () => {
        const response = await request.get('https://jsonplaceholder.typicode.com/todos/1').set({
          'Accept-Encoding': 'identity' // This will prevent responses from being compressed which is preferred for mock recording
        })
        await exporter.exportResponseBody('todoMock', response)
        await exporter.exportResponseBody('todoMock', response)
        const importedMock = await import(getMockPath('todoMock.ts'))
        expect(importedMock).toMatchSnapshot()
        const importedMock2 = await import(getMockPath('todoMock_1.ts'))
        expect(importedMock2).toMatchSnapshot()
      })
    })

    it('Overwrites mocks when configured', async () => {
      rimraf.sync(getMockPath('todoMockToRewrite.ts'))
      const helper = initHelper(__dirname)

      await helper.nockBack('placeholder-GET.json', async () => {
        const response = await request.get('https://jsonplaceholder.typicode.com/todos/1').set({
          'Accept-Encoding': 'identity' // This will prevent responses from being compressed which is preferred for mock recording
        })

        const overwritingExporter = new ResponseTypescriptExporter(__dirname + '/nock-responses', {
          doNotOverwrite: false
        })
        await overwritingExporter.exportFullResponse('todoMockToRewrite', response)
        await overwritingExporter.exportResponseBody('todoMockToRewrite', response)
        const importedMock = await import(getMockPath('todoMockToRewrite.ts'))
        expect(importedMock).toMatchSnapshot()
      })
    })
  })
})

function getMockPath(fixtureName: string) {
  return __dirname + '/nock-responses/' + fixtureName
}
