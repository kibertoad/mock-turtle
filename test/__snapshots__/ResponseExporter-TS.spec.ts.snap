// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResponseExporter Export to TypeScript Does not overwrite mocks by default 1`] = `
Object {
  "ResponseMock": Object {
    "completed": false,
    "id": 1,
    "title": "delectus aut autem",
    "userId": 1,
  },
}
`;

exports[`ResponseExporter Export to TypeScript Does not overwrite mocks by default 2`] = `
Object {
  "ResponseMock": Object {
    "completed": false,
    "id": 1,
    "title": "delectus aut autem",
    "userId": 1,
  },
}
`;

exports[`ResponseExporter Export to TypeScript Export full response 1`] = `
Object {
  "ResponseMock": Object {
    "body": Object {
      "completed": false,
      "id": 1,
      "title": "delectus aut autem",
      "userId": 1,
    },
    "headers": Object {
      "accept-ranges": "bytes",
      "access-control-allow-credentials": "true",
      "age": "4207",
      "cache-control": "public, max-age=14400",
      "cf-cache-status": "HIT",
      "cf-ray": "50ce2884ef319c3f-AMS",
      "connection": "close",
      "content-length": "83",
      "content-type": "application/json; charset=utf-8",
      "date": "Tue, 27 Aug 2019 12:48:11 GMT",
      "etag": "W/\\"53-hfEnumeNh6YirfjyjaujcOPPT+s\\"",
      "expect-ct": "max-age=604800, report-uri=\\"https://report-uri.cloudflare.com/cdn-cgi/beacon/expect-ct\\"",
      "expires": "Tue, 27 Aug 2019 16:48:11 GMT",
      "pragma": "no-cache",
      "server": "cloudflare",
      "set-cookie": "__cfduid=d6bfb4cafb2fadd8a6ab89a6195a1a61c1566910091; expires=Wed, 26-Aug-20 12:48:11 GMT; path=/; domain=.typicode.com; HttpOnly",
      "vary": "Origin, Accept-Encoding",
      "via": "1.1 vegur",
      "x-content-type-options": "nosniff",
      "x-powered-by": "Express",
    },
    "responseCode": 200,
  },
}
`;

exports[`ResponseExporter Export to TypeScript Export only body 1`] = `
Object {
  "ResponseMock": Object {
    "completed": false,
    "id": 1,
    "title": "delectus aut autem",
    "userId": 1,
  },
}
`;

exports[`ResponseExporter Export to TypeScript Overwrites mocks when configured 1`] = `
Object {
  "ResponseMock": Object {
    "completed": false,
    "id": 1,
    "title": "delectus aut autem",
    "userId": 1,
  },
}
`;
